File: models1.py
"""
supervised models for dataset 1
Logistic regression and random forest
There are two runs, because the dataset1 has two excel sheets. It was suggested
to use the tempdata as a test set. Because I loose especially the DTI parameter
I also trained models on only the newdata set.
Thats why the code is split into 1st evaluation and 2nd evaluation
"""

import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix, roc_curve, auc

plt.rcParams.update(
    {
        "font.size": 14,
        "axes.titlesize": 16,
        "axes.labelsize": 14,
        "xtick.labelsize": 14,
        "ytick.labelsize": 14,
        "legend.fontsize": 12,
        "figure.titlesize": 18,
    }
)
plt.rcParams["figure.figsize"] = (8, 5)

df = pd.read_excel("data/Dataset_Assignment_1.xlsx", sheet_name=0)
temp_data = pd.read_excel("data/Dataset_Assignment_1.xlsx", sheet_name=1)

# 1st evaluation: Training data and test data from newdata
X = df.drop(columns=["ID", "Decision"])
X = pd.get_dummies(X, columns=["Gender", "Accommodation_Class"], drop_first=True)
y = df["Decision"].map({"Reject": 0, "Approve": 1})

X_train1, X_test1, y_train1, y_test1 = train_test_split(
    X, y, test_size=0.3, random_state=123, stratify=y
)
print(f"Number of Training Data {X_train1.shape[0]}")
print(f"Number of Test Data {y_test1.shape[0]}")
ratio1 = y_train1.value_counts(normalize=True) * 100
print("Ration of Approve/Reject in Training Data: ", ratio1)

# 2nd evaluation: Training data -> newdata , Test data -> tempdata
X_train2 = df.drop(columns=["ID", "DTI", "Gender", "Accommodation_Class", "Decision"])
y_train2 = df["Decision"].map({"Reject": 0, "Approve": 1})

X_test2 = temp_data.drop(
    columns=["Applicant ID", "Tier of City", "Years in Address", "Decision"]
)
X_test2.rename(columns={"Old EMI amount": "OldEmi"}, inplace=True)
X_test2 = X_test2[X_train2.columns]
y_test2 = temp_data["Decision"].map({"Reject": 0, "Approve": 1})
print(f"Number of Training Data {X_train2.shape[0]}")
print(f"Number of Test Data {y_test2.shape[0]}")
ratio2 = y_train2.value_counts(normalize=True) * 100
print("Ration of Approve/Reject in Training Data: ", ratio2)


def run_model(X_train, X_test, y_train, y_test, run):
    """
    runs the logistic regression and random forest

    Args:
        X_train (_type_): _description_
        X_test (_type_): _description_
        y_train (_type_): _description_
        y_test (_type_): _description_
    """
    scaler = StandardScaler()
    X_train_scaled = scaler.fit_transform(X_train)
    X_test_scaled = scaler.transform(X_test)

    # Logistic Regression model
    log_model = LogisticRegression(
        random_state=123, class_weight="balanced", max_iter=1000
    )
    log_model.fit(X_train_scaled, y_train)
    y_prob_log = log_model.predict_proba(X_test_scaled)[:, 1]
    threshold_log = 0.5
    y_pred_log = (y_prob_log >= threshold_log).astype(int)
    # Random Forest model
    rf_model = RandomForestClassifier(
        n_estimators=100, class_weight="balanced", random_state=123
    )
    rf_model.fit(X_train, y_train)  # not scaled
    y_prob_rf = rf_model.predict_proba(X_test)[:, 1]
    threshold_rf = 0.5
    y_pred_rf = (y_prob_rf >= threshold_rf).astype(int)

    def conf_matrix(y_pred, model):
        cm = confusion_matrix(y_test, y_pred)
        tn = cm[0, 0]
        fp = cm[0, 1]
        fn = cm[1, 0]
        tp = cm[1, 1]
        accuracy = (tp + tn) / (tp + tn + fp + fn)
        precision = tp / (tp + fp)
        recall = tp / (tp + fn)
        f1 = 2 * (precision * recall) / (precision + recall)
        print(f"------{model}---------")
        print(cm)
        print(f"Accuracy: {accuracy*100:.2f}%")
        print(f"Precision: {precision*100:.2f}%")
        print(f"Recall (TPR): {recall*100:.2f}%")
        print(f"F1 Score: {f1*100:.2f}%")
        print("---------------")

    conf_matrix(y_pred_log, "Logistic Regression")
    conf_matrix(y_pred_rf, "Random Forest")

    fpr_log, tpr_log, thresholds_log = roc_curve(y_test, y_prob_log)
    fpr_rf, tpr_rf, thresholds_rf = roc_curve(y_test, y_prob_rf)
    plt.figure(figsize=(6, 6))
    plt.plot(fpr_log, tpr_log, label="Logistic Regression")
    plt.plot(fpr_rf, tpr_rf, label="Random Forest")
    plt.plot()
    plt.plot([0, 1], [0, 1], "--", color="gray", label="Random classifier")
    plt.xlabel("False Positive Rate")
    plt.ylabel("True Positive Rate")
    plt.legend()
    plt.tight_layout()
    plt.savefig(f"pictures/roc_run{run}.pdf", bbox_inches="tight")
    plt.show()

    def benchmark(fpr, tpr, model):
        print(f"------{model}---------")
        auc_val = auc(fpr, tpr)
        print(f"AUC: {auc_val:.2f}")
        gini = 2 * auc_val - 1
        print(f"Gini: {gini:.2f}")
        print("---------------")

    benchmark(fpr_log, tpr_log, "Logistic Regressiono")
    benchmark(fpr_rf, tpr_rf, "Random Forest")

    # additional random forest analysis
    feature_names = X_train.columns.tolist()
    feature_names = [
        (
            "Gender"
            if col.startswith("Gender")
            else "AC" if col.startswith("Accommodation_Class") else col
        )
        for col in feature_names
    ]
    importance = pd.Series(
        rf_model.feature_importances_, index=feature_names
    ).sort_values(ascending=False)
    print("Importance Random forest")
    print(importance)
    importance.plot(kind="bar")
    plt.ylabel("Importance")
    plt.savefig(f"pictures/importantce_run{run}.pdf", bbox_inches="tight")
    plt.show()


print("-----------1st experiment-----------")
run_model(X_train1, X_test1, y_train1, y_test1, run="1")
print("-----------------------")
print("-----------2nd experiment-----------")
run_model(X_train2, X_test2, y_train2, y_test2, run="2")
print("-----------------------")

File: models1.R
# =====================================================
# Supervised models for Dataset 1
# Logistic Regression and Random Forest
# =====================================================

library(readxl)
library(caret)
library(randomForest)
library(pROC)

set.seed(123)

# -----------------------------------------------------
# Load data
# -----------------------------------------------------
getwd()
df <- read_excel("data/Dataset_Assignment_1.xlsx", sheet = 1)
temp_data <- read_excel("data/Dataset_Assignment_1.xlsx", sheet = 2)

# -----------------------------------------------------
# 1st evaluation
# Random train/test split
# -----------------------------------------------------

X <- subset(df, select = -c(ID, Decision))

# Convert categorical variables to dummy variables
X <- model.matrix(~ . - 1, data = X)

y <- factor(df$Decision,
            levels = c("Reject", "Approve"))

train_index <- createDataPartition(y,
                                   p = 0.7,
                                   list = FALSE)

X_train1 <- X[train_index, ]
X_test1  <- X[-train_index, ]

y_train1 <- y[train_index]
y_test1  <- y[-train_index]

cat("Training observations:", nrow(X_train1), "\n")
cat("Test observations:", nrow(X_test1), "\n")

prop.table(table(y_train1))

# -----------------------------------------------------
# 2nd evaluation
# Train on Dataset 1
# Test on temporary dataset
# -----------------------------------------------------

X_train2 <- subset(df,
                   select = -c(ID,
                               DTI,
                               Gender,
                               Accommodation_Class,
                               Decision))

y_train2 <- factor(df$Decision,
                   levels = c("Reject","Approve"))

X_test2 <- subset(temp_data,
                  select = -c(`Applicant ID`,
                              `Tier of City`,
                              `Years in Address`,
                              Decision))

names(X_test2)[names(X_test2)=="Old EMI amount"] <- "OldEmi"

X_test2 <- X_test2[, names(X_train2)]

y_test2 <- factor(temp_data$Decision,
                  levels = c("Reject","Approve"))

cat("Training observations:", nrow(X_train2), "\n")
cat("Test observations:", nrow(X_test2), "\n")

prop.table(table(y_train2))

# =====================================================
# Function to run both models
# =====================================================

run_model <- function(X_train,
                      X_test,
                      y_train,
                      y_test,
                      run){
  
  # ---------------------------------------------------
  # Standardization
  # ---------------------------------------------------
  
  scaler <- preProcess(X_train,
                       method = c("center","scale"))
  
  X_train_scaled <- predict(scaler, X_train)
  X_test_scaled <- predict(scaler, X_test)
  
  # ---------------------------------------------------
  # Logistic Regression
  # ---------------------------------------------------
  
  train_log <- data.frame(X_train_scaled,
                          Decision = y_train)
  
  log_model <- glm(
    Decision ~ .,
    data = train_log,
    family = binomial()
  )
  
  prob_log <- predict(
    log_model,
    newdata = X_test_scaled,
    type = "response"
  )
  
  pred_log <- ifelse(prob_log >= 0.5,
                     "Approve",
                     "Reject")
  
  pred_log <- factor(pred_log,
                     levels = levels(y_test))
  
  # ---------------------------------------------------
  # Random Forest
  # ---------------------------------------------------
  
  train_rf <- data.frame(X_train,
                         Decision = y_train)
  
  rf_model <- randomForest(
    Decision ~ .,
    data = train_rf,
    ntree = 100,
    importance = TRUE
  )
  
  prob_rf <- predict(
    rf_model,
    newdata = X_test,
    type = "prob")[,"Approve"]
  
  pred_rf <- ifelse(prob_rf >= 0.5,
                    "Approve",
                    "Reject")
  
  pred_rf <- factor(pred_rf,
                    levels = levels(y_test))
  
  # ---------------------------------------------------
  # Performance metrics
  # ---------------------------------------------------
  
  evaluate_model <- function(pred,
                             truth,
                             model_name){
    
    cm <- confusionMatrix(pred,
                          truth,
                          positive = "Approve")
    
    cat("\n----------------------------\n")
    cat(model_name,"\n")
    cat("----------------------------\n")
    
    print(cm$table)
    
    cat("Accuracy:",
        round(cm$overall["Accuracy"],3),"\n")
    
    cat("Precision:",
        round(cm$byClass["Pos Pred Value"],3),"\n")
    
    cat("Recall:",
        round(cm$byClass["Sensitivity"],3),"\n")
    
    cat("F1:",
        round(cm$byClass["F1"],3),"\n")
    
  }
  
  evaluate_model(pred_log,
                 y_test,
                 "Logistic Regression")
  
  evaluate_model(pred_rf,
                 y_test,
                 "Random Forest")
  
  # ---------------------------------------------------
  # ROC curves
  # ---------------------------------------------------
  
  roc_log <- roc(y_test,
                 prob_log)
  
  roc_rf <- roc(y_test,
                prob_rf)
  
  plot(roc_log,
       col = "blue",
       main = "ROC Curve")
  
  lines(roc_rf,
        col = "red")
  
  legend("bottomright",
         legend = c("Logistic Regression",
                    "Random Forest"),
         col = c("blue","red"),
         lwd = 2)
  
  # ---------------------------------------------------
  # AUC / Gini
  # ---------------------------------------------------
  
  cat("\nLogistic Regression\n")
  cat("AUC:", auc(roc_log), "\n")
  cat("Gini:", 2*auc(roc_log)-1,"\n")
  
  cat("\nRandom Forest\n")
  cat("AUC:", auc(roc_rf), "\n")
  cat("Gini:", 2*auc(roc_rf)-1,"\n")
  
  # ---------------------------------------------------
  # Variable Importance
  # ---------------------------------------------------
  
  importance <- importance(rf_model)
  
  print(importance)
  
  varImpPlot(rf_model)
  
}

# =====================================================
# Run experiments
# =====================================================

cat("\n=========== Experiment 1 ===========\n")

run_model(
  X_train1,
  as.data.frame(X_test1),
  y_train1,
  y_test1,
  run = 1
)

cat("\n=========== Experiment 2 ===========\n")

run_model(
  X_train2,
  X_test2,
  y_train2,
  y_test2,
  run = 2
)

